Support HTTPS connection to Jupyter notebook

danielhollas opened the pull request:
Basic aiidalab container comes with HTTP, but users might want to add HTTPS.
Here we just modify the code that tries to connect to Jupyter Notebook to determine whether it is ready.

I am currently playing with HTTPS for local development in our Docker image built on top of AiiDAlab-docker-stack, see https://github.com/danielhollas/aiidalab-ispg-docker-stack/pull/6 if that is of interest.

**TODO**: 
- [ ] Tests. Seems like it would be tricky to test this properly, we would somehow need to patch the Jupyter notebook in a  running container? Not sure it's worth investing too much time. In any case, to enable HTTPS, one simply needs to pass the key and certificate to `jupyter-notebook` in `/opt/start-notebook.sh`, see the PR linked above.
- [x] When the start command finishes, we still print http url, and if user tries to open the browser directly from cmdline it will open http and fail. I didn't figure out how to improve this since the polling function is an async `@staticmethod` so I did not not how to cleanly pass info from it without larger refactor.

csadorf commented:
The aiidalab-launch tool is designed for local invocation not necessarily for serving the service via a proxy over the wider network. I must admit that I am having a bit of a hard time following the use case, maybe you can elaborate a little bit?

Either way, I have no problem with accepting a slight patch to the tool if it unblocks the effort as long as it does not increase the complexity disproportionally.

Is this PR ready for review or comment or are you still experimenting? In the latter case I'd suggest to change it into draft stage, otherwise, please request my review and I will have a closer look at the implementation.

danielhollas commented:
@csadorf this is ready for review.

Well, having https in local environment is good for testing it out, and also when using `mkcert` and issuing your own root certificate you can completely get away from the annoying browser warning when you're accessing http site (on Firefox I am using the strict only-http mode).

Besided local development, I have another server where I am essentially serving Aiidalab application within our VPN for members of our group. Here again having HTTPS is very much desirable. Also I am not yet ready to dive into full-blown production deployment so using `aiidalab-launch` makes sense for me. Let me know if I can clarify further.

csadorf reviewed (changes requested):
It looks this PR introduces a few unrelated changes to the logging behavior and verbosity. Would you mind removing those or alternatively justify why they should be part of this change set? Thank you!

danielhollas commented on aiidalab_launch/__main__.py:
```diff
@@ -415,9 +415,11 @@ async def _async_start(
-                    "verbosity with 'aiidalab-launch -vvv start'."
+                    "The AiiDAlab instance failed to start. You can inspect "
+                    "the container output logs with "
+                    f"'aiidalab-launch logs -p {instance.profile.name}' "
```
In my experience this is a better advice in most cases.

danielhollas commented on aiidalab_launch/instance.py:
```diff
@@ -286,34 +286,45 @@ async def echo_logs(self) -> None:
     async def _init_scripts_finished(container: Container) -> None:
         loop = asyncio.get_event_loop()
-        LOGGER.debug("Waiting for init services to finish...")
+        LOGGER.info("Waiting for init services to finish...")
```
I changed a few imo useful debug statements into info. Main reason is that I find that '-vv' does not print enough, while '-vvv' prints way too much, mostly due to notebook polling. Happy to reverse this though.

danielhollas commented:
Yes, sorry about that. I added justification in comments but am happy to submit those in separate PR. 

However, there is no "functional" change to logging behaviour, I just changed a logging level of a couple of statement to make the '-vv' level more useful.

csadorf commented:
> Yes, sorry about that. I added justification in comments but am happy to submit those in separate PR.
> 
> However, there is no "functional" change to logging behaviour, I just changed a logging level of a couple of statement to make the '-vv' level more useful.

Ok, fair enough. Feel free to leave the adjustments in the PR.

csadorf commented on aiidalab_launch/instance.py:
```diff
@@ -286,34 +286,45 @@ async def echo_logs(self) -> None:
         loop = asyncio.get_event_loop()
-        LOGGER.debug("Waiting for notebook service to become reachable...")
+        LOGGER.info("Waiting for notebook service to become reachable...")
+        proto = "http"
```
Can we give this variable a slightly more descriptive name? Maybe something like:
```suggestion
        assumed_protocol = "http"
```

csadorf commented on aiidalab_launch/instance.py:
```diff
@@ -286,34 +286,45 @@ async def echo_logs(self) -> None:
                 elif result.exit_code in (7, 28):
                     await asyncio.sleep(2)  # jupyter not yet reachable
                     continue
+                elif result.exit_code == 56 and proto == "http":
+                    proto = "https"
+                    LOGGER.info("Trying to connect via HTTPS.")
+                    continue
```
According to the curl manual, exit code 56 indicates a "[f]ailure in receiving network data." To my understanding this _can_ be caused by the wrong protocol, but there could of course also be other reasons for this failure mode. How confident are you that switching the protocol at this point is _always_ the right move?

Should we maybe try both protocols?

danielhollas replied:
I am not sure what you're proposing. 

> How confident are you that switching the protocol at this point is always the right move?
Not confident at all. I was hoping that curl will give me a more specific error when trying the wrong protocol, but alas this is what I am getting in Ubuntu 20.04 and curl 7.68. Note however that this PR does not change behavior with respect to the http protocol. If we get this error code, we simply try https once and then bail out if that doesn't work (so in the worst case we just added an extra useless ping).

csadorf replied:
Yes, I misread the logic earlier and thought the protocol was essentially permanently switched. I think this is good like this. 👍

danielhollas replied:
Great, thanks for the review! I've addressed the other bits.

csadorf commented on aiidalab_launch/instance.py:
```diff
@@ -286,34 +286,45 @@ async def echo_logs(self) -> None:
+                elif result.exit_code == 60:
+                    # This is fine for local development
+                    LOGGER.warn(
+                        "Could not authenticate HTTPS certificate of Jupyter notebook"
```
```suggestion
                        "Could not authenticate HTTPS certificate."
```

csadorf commented on aiidalab_launch/instance.py:
```diff
@@ -286,34 +286,45 @@ async def echo_logs(self) -> None:
+                    LOGGER.info("Trying to connect via HTTPS.")
+                    continue
+                elif result.exit_code == 60:
+                    # This is fine for local development
```
I'm not sure that this comment adds much value. Whether this is fine or not is till up to the judgement of the developer/user.

csadorf commented:
> * [ ]  Tests. Seems like it would be tricky to test this properly, we would somehow need to patch the Jupyter notebook in a  running container? Not sure it's worth investing too much time. In any case, to enable HTTPS, one simply needs to pass the key and certificate to `jupyter-notebook` in `/opt/start-notebook.sh`, see the PR linked above.

You wouldn't need to patch a running container, we could adjust it before the start. But I agree, this is esoteric enough that the effort is likely not justified. If you come up with a good solution to test it, I'd of course welcome it. 

> * [ ]  When the start command finishes, we still print http url, and if user tries to open the browser directly from cmdline it will open http and fail. I didn't figure out how to improve this since the polling function is an async `@staticmethod` so I did not not how to cleanly pass info from it without larger refactor.

I'd suggest to simply return it and then toggle an attribute on the `AiidaLabInstance` class. If you like I can implement that, just let me now.

danielhollas commented:
> I'd suggest to simply return it and then toggle an attribute on the AiidaLabInstance class. If you like I can implement that, just let me now.

I didn't know how to return from an async function, but it was quite easy in the end. :smile: We now print the correct protocol both at the and of the start command and also in the output of the status command. :tada: 

![image](https://user-images.githubusercontent.com/9539441/181608430-4528da29-c12f-4a19-a65b-18e29e22833b.png)

csadorf commented on aiidalab_launch/instance.py:
```diff
@@ -65,6 +65,7 @@ class AiidaLabInstanceStatus(Enum):
     profile: Profile
     _image: docker.models.images.Image = None
     _container: Container = None
+    _protocol: str = "http"
```
I'm wondering whether the default value should be `None` instead. Basically we do not know what the protocol is unless the container was started. In that case we should reset this back to `None` when we stop the container. What do you think?

danielhollas replied:
Hmm, not sure this is worth the extra complexity, then we would need to special-case the url() function.

csadorf replied:
Ok, fair enough.

csadorf commented:
@danielhollas Thanks for the contribution! I just released this feature in https://github.com/aiidalab/aiidalab-launch/releases/tag/v2022.1015 .

danielhollas commented:
Thanks so much @csadorf ! : 💖
